### PR TITLE
Fix S1905 FP: Nullability context and array of anonymous types

### LINKED ISSUE #6438: Fix S1905 FP: Nullability context and array of anonymous types
### Description

In a .net 6 project where nullable reference types is enabled, in the example below the casts to string? are necessary, otherwise the nullability of x in the two array elements doesn't match and this produces a  CS8206 error 

### Repro steps

````
var anon = new[] { new { x = (string?)"foo" }, new { x = (string?)null } };
````

### Expected behavior

No S1905 warning is produced

### Actual behavior

S1905 warning is produced

### Known workarounds

Don't use anonymous types and create a class with a property x of type string? and declare the array of that type

### Related information

* C# 8 with nullable reference types on
* VS 2022 19.4.0
* SonarScanner 8.49.0.57237
* Operating System Windows Server 2022

### PR DESCRIPTION
**Pull request by Tim-Pohlmann**
Fixes #6438

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs [hunk 1]
```diff
@@ -18,161 +18,167 @@
...
+                ? memberAccess.GetLocation()
+                : memberAccess.OperatorToken.CreateLocation(invocation),
+            _ => invocation.Expression.GetLocation()
+        };
+
+    private static ITypeSymbol GetElementType(InvocationExpressionSyntax invocation,
+                                              IMethodSymbol methodSymbol,
+                                              SemanticModel semanticModel)
+    {
+        ExpressionSyntax collection;
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs:158]: Should be rewritten about like this
```cs
        return CollectionExpression() is { } collection
            ? semanticModel.GetTypeInfo(collection).Type switch
            {
                INamedTypeSymbol { TypeArguments: { Length: 1 } typeArguments } => typeArguments.First(),
                IArrayTypeSymbol { Rank: 1 } arrayType => arrayType.ElementType,    // casting is necessary for multidimensional arrays
                _ => null
            }
            : null;

        ExpressionSyntax CollectionExpression()
        {
            if (methodSymbol.MethodKind == MethodKind.Ordinary)
            {
                if (invocation is { ArgumentList.Arguments: { Count: > 0 } arguments }
                    && arguments.First() is { Expression: { } firstArgumentExpression })
                {
                    return firstArgumentExpression;
                }
            }
            else if (invocation is { Expression: MemberAccessExpressionSyntax { Expression: { } memberAccessExpression } })
            {
                return memberAccessExpression;
            }

            return null;
        }
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/RedundantCastTest.cs [hunk 2]
```diff
@@ -21,45 +21,99 @@
...
+    public void RedundantCast_NullableAnnotations(string snippet, bool compliant)
+        => VerifyNullableTests(snippet, "enable annotations", compliant);
+
+    private void VerifyNullableTests(string snippet, string nullableContext, bool compliant)
+    {
+        var code = $$"""
+            #nullable {{nullableContext}}
+            void Test(string nonNullable, string? nullable)
+            {
+                {{snippet}} // {{compliant switch { true => "Compliant", false => "Noncompliant" }}}
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/RedundantCastTest.cs:114]: NIT: I personally prefer a ternary in such simple cases.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/RedundantCastTest.cs [hunk 3]
```diff
@@ -21,45 +21,99 @@
...
+        new object[] { """_ = (string)nonNullable;""", false, false },
+        new object[] { """_ = (string?)nonNullable;""", true, false },
+        new object[] { """if (nullable != null) _ = (string)nullable;""", false, false },
+        new object[] { """if (nullable != null) _ = (string?)nullable;""", true, false },
+        new object[] { """if (nonNullable == null) _ = (string)nonNullable;""", true, false },
+        new object[] { """if (nonNullable == null) _ = (string?)nonNullable;""", false, false },
+    };
+
+    private static IEnumerable<object[]> NullableTestDataWithFlowState =>
+        NullableTestData.Select(x => new[] { x[0], x[1] });
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/RedundantCastTest.cs:83]: If you always do the select, you can also strong type the return type of `NullableTestData`:
```cs
private static IEnumerable<(string snippet, bool compliantWithFlowState, bool compliantWithoutFlowState)> NullableTestData()
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs [hunk 4]
```diff
@@ -2,15 +2,50 @@
...
+        {
+            var anon = new { X = (string?)"foo" };  // Compliant
+        }
+
+        public void Array()
+        {
+            var anonArray = new[] { new { X = (string?)"foo" }, new { X = (string?)null } };    // Compliant
+            var oneElementAnonArray = new[] { new { X = (string?)"foo" } };                     // Compliant
+            var notSoAnonArray = new[] { new HoldsObject(new { X = (string?)"foo" }) };         // Compliant
+        }
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs:31]: Please add tests for tuple construction:
```cs
_ = (a: (string?)"", b: "");
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs [hunk 5]
```diff
@@ -2,15 +2,50 @@
...
 namespace Tests.Diagnostics
 {
     // https://github.com/SonarSource/sonar-dotnet/issues/3273
     public class CastOnNullable
     {
-        public static IEnumerable<string> UsefulCast()
+        public static IEnumerable<string> Array()
         {
             var nullableStrings = new string?[] { "one", "two", null, "three" };
             return nullableStrings.OfType<string>(); // Compliant - filters out the null
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs:14]: Did you have a look if we can merge our new logic with the one that detects this case?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs [hunk 6]
```diff
@@ -18,161 +18,167 @@
...
+                    return;
                 }
-                collection = invocation.ArgumentList.Arguments.First().Expression;
-            }
-            else
-            {
-                if (!(invocation.Expression is MemberAccessExpressionSyntax memberAccess))
+
+                var elementType = GetElementType(invocation, methodSymbol, context.SemanticModel);
+                // Generic types {T} and {T?} are equal and there is no way to access NullableAnnotation field right now
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs:112]: I don't think this is true anymore. You may want to have a look here as well (if it turns out to be too complicated we may also create an issue and come back to it later).

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs [hunk 7]
```diff
@@ -18,161 +18,167 @@
...
-                : memberAccess.OperatorToken.CreateLocation(invocation);
+        if (expressionType.Equals(castType) && FlowStateIsNoneOrMatchesCast(expressionTypeInfo, type))
+        {
+            ReportIssue(context, expression, location, castType);
         }
+    }
+
+    private static bool FlowStateIsNoneOrMatchesCast(TypeInfo expressionTypeInfo, ExpressionSyntax type)
+    {
+        var castingToNullable = type is NullableTypeSyntax;
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs:87]: NIT: `NullableTypeSyntax` is sealed. It is also possible to use `type.IsKind(..)` therefore which is slightly better.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs [hunk 8]
```diff
@@ -18,161 +18,167 @@
...
-            if (expressionType.Equals(castType))
-            {
-                context.ReportIssue(Diagnostic.Create(rule, location,
-                    castType.ToMinimalDisplayString(context.SemanticModel, expression.SpanStart)));
-            }
+    private static void CheckCastExpression(SonarSyntaxNodeReportingContext context, ExpressionSyntax expression, ExpressionSyntax type, Location location)
+    {
+        if (expression.IsKind(SyntaxKindEx.DefaultLiteralExpression))
+        {
+            return;
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs:64]: We do not have tests for the `as` expression call to CheckCastExpression in conjunction with FlowState. There are probably other rules at play. Can you add some tests?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs [hunk 9]
```diff
@@ -18,161 +18,167 @@
...
+        && methodSymbol.IsExtensionOn(KnownType.System_Collections_IEnumerable)
+            ? methodSymbol
+            : null;
+
+    private static ITypeSymbol GetGenericTypeArgument(ITypeSymbol type) =>
+        type is INamedTypeSymbol returnType && returnType.Is(KnownType.System_Collections_Generic_IEnumerable_T)
+            ? returnType.TypeArguments.Single()
+            : null;
+
+    private static bool CanHaveNullValue(ITypeSymbol type) => type.IsReferenceType || type.Name == "Nullable";
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs:143]: Formatting is of here
```suggestion
    private static bool CanHaveNullValue(ITypeSymbol type) =>
        type.IsReferenceType || type.Name == "Nullable";
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Some nice refactoring and the solution seems also right to me.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs (hunk 6 above)
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs:99]: I don't see a straight-forward solution. Let's come back to it later.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs (hunk 5 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs:14]: not straight forward

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs (hunk 1 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs:141]: This change is missing

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs [hunk 10]
```diff
@@ -18,161 +18,150 @@
...
-            return methodCalledAsStatic
+    private static Location GetReportLocation(InvocationExpressionSyntax invocation, bool methodCalledAsStatic) =>
+        invocation.Expression switch
+        {
+            MemberAccessExpressionSyntax memberAccess => methodCalledAsStatic
                 ? memberAccess.GetLocation()
-                : memberAccess.OperatorToken.CreateLocation(invocation);
-        }
+                : memberAccess.OperatorToken.CreateLocation(invocation),
+            _ => invocation.Expression.GetLocation()
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs:136]: This case is not covered. You may be able to cover the case by using other access kinds: [sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzguATEBqAPgAQCYAGAWACgiBGC/YgAnyoDoAZWARwG5aGmA2Ri3YwOzAKIwArmACmAJwjAANrJ7laVAJwAKAESAeDcAg+3oCU6gG4R59a4oCe9ALz0YsgO71YCANoBdAG8AX3Vvf39beQcAeQAzAEEoiEcXN08w4j9/dQoAfWdIhwB+ZgBhVAQAHm8APh1zegB6RvoAWVkwYAUAIVgMWABzPIK7ZLjEhx9MsorqmAQ6hub6cVU5efiAY03ZFBRhl1H7caT7Iqm/GbQ5hfquJpbVjtl53ph+mAH6HQB7AGtkq4fgh6PJZBBlMpHD8YPRymg4PRgFIQSgfnJ3AALBSyeiyZQoWSmA7fAAiEAQsgAKlA5MwAHI/dzMckpFxUehFQrJACE9BArg8XnmF1MVyqtTuD3olPkMGs9iAA=)

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs (hunk 5 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs:14]: Please create a ticket about lighting up the nullability info on ITypeSymbol on the shim layer and link to https://github.com/SonarSource/sonar-dotnet/issues/3273#issuecomment-606693989 and this comment thread.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs (hunk 1 above)
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs]: Simplified a bit further.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs (hunk 5 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs:14]: I'm not super clear on what that means and thus don't know what to put in the ticket exactly.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/RedundantCastTest.cs [hunk 11]
```diff
@@ -21,45 +21,104 @@
...
+        ("""_ = (string)"Test";""", false, false),
+        ("""_ = (string?)"Test";""", true, false),
+        ("""_ = (string)null;""", true, true),
+        ("""_ = (string?)null;""", true, true),
+        ("""_ = (string)nullable;""", true, false),
+        ("""_ = (string?)nullable;""", false, false),
+        ("""_ = (string)nonNullable;""", false, false),
+        ("""_ = (string?)nonNullable;""", true, false),
+        ("""_ = nullable as string;""", true, false),
+        ("""_ = nonNullable as string;""", false, false),
```
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/RedundantCastTest.cs:78]: This does not compile. Nullable reference types cannot be used in an "as"-expression.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/RedundantCastTest.cs [hunk 12]
```diff
@@ -21,45 +21,104 @@
...
+        ("""if (nullable != null) _ = (string)nullable;""", false, false),
+        ("""if (nullable != null) _ = (string?)nullable;""", true, false),
+        ("""if (nullable != null) _ = nullable as string;""", false, false),
```
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/RedundantCastTest.cs:81]: Does not compile.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/RedundantCastTest.cs [hunk 13]
```diff
@@ -21,45 +21,104 @@
...
+        ("""if (nonNullable == null) _ = (string)nonNullable;""", true, false),
+        ("""if (nonNullable == null) _ = (string?)nonNullable;""", false, false),
+        ("""if (nonNullable == null) _ = nonNullable as string;""", true, false),
```
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/RedundantCastTest.cs:84]: Does not compile.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs [hunk 14]
```diff
@@ -2,15 +2,63 @@
 using System.Collections.Generic;
 using System.Linq;
 
+#nullable enable
```
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs:5]: For some reason this fails with:
`Unexpected build error [CS1040]: Preprocessor directives must appear as the first non-whitespace character on a line on line 1`

[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs:5]: This is due to the "concurrent" analysis in the tests. The whole file gets wrapped in its own namespace and is analyzed in parallel. You can turn this of in the VerifierBuilder via `WithConcurrentAnalysis`.
You can leave the nullable enable here until we fix the issue in the verifier (see e.g. [#6211](https://github.com/SonarSource/sonar-dotnet/pull/6211))

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs [hunk 15]
```diff
@@ -18,158 +18,147 @@
...
+            : null;
 
-            return methodCalledAsStatic
-                ? memberAccess.GetLocation()
-                : memberAccess.OperatorToken.CreateLocation(invocation);
-        }
-
-        private static ITypeSymbol GetElementType(InvocationExpressionSyntax invocation, IMethodSymbol methodSymbol,
-            SemanticModel semanticModel)
+        ExpressionSyntax CollectionExpression()
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs:150]: Coverage is bad in this method. You probably need some non-compiling code to increase coverage here. It tests "Cast" and "TypeOf" for different kinds of invocations. You may try
```cs
col?.Cast<int>(); // MemberBinding
colOfcol[0].Cast<int>(); // ElementAccess
colOfcol[0]?.Cast<int>(); // ElementBinding
func().Cast<int>(); // func is Func<IEnumerable<int>>
func()?.Cast<int>()?.Where(x=>true);
Enumerable.Cast<int>(); // compiler error (test for FirstAndDefault in the unreduced extension method call path)
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs [hunk 16]
```diff
@@ -18,158 +18,147 @@
...
-            var typeInfo = semanticModel.GetTypeInfo(collection);
-            if (typeInfo.Type is INamedTypeSymbol collectionType &&
-                collectionType.TypeArguments.Length == 1)
-            {
-                return collectionType.TypeArguments.First();
-            }
-
-            if (typeInfo.Type is IArrayTypeSymbol arrayType &&
-                arrayType.Rank == 1) // casting is necessary for multidimensional arrays
+            else if (invocation is { Expression: MemberAccessExpressionSyntax { Expression: { } memberAccessExpression } })
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs:159]: `GetParentConditionalAccessExpression` should help to fix some of the issues I described in the coverage comment above.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Can you bump the coverage in the touched code paths?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs (hunk 5 above)
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/RedundantCast.CSharp8.cs:14]: https://github.com/SonarSource/sonar-dotnet/issues/6825

**Comment by Tim-Pohlmann** [comment]
When we have `GetLeftSideOfDot` from Roslyn (see https://github.com/SonarSource/sonar-dotnet/issues/6863), it should be used in the implementation of this rule.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs [hunk 17]
```diff
@@ -18,161 +18,139 @@
...
-
-            if (typeInfo.Type is IArrayTypeSymbol arrayType &&
-                arrayType.Rank == 1) // casting is necessary for multidimensional arrays
-            {
-                return arrayType.ElementType;
-            }
-
-            return null;
-        }
+                : invocation.ArgumentList.Arguments.FirstOrDefault()?.Expression;
```
Comment (Tim-Pohlmann) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs:154]: When we have `GetLeftSideOfDot` from Roslyn (see https://github.com/SonarSource/sonar-dotnet/issues/6863), it should be used to refactor this.

**Comment by Tim-Pohlmann** [comment]
@pavel-mikula-sonarsource can you merge this? lack of coverage is due to 3 lines of defensive programming.

**Pull request merged**